src/biz/collab/utils.rs

```diff
@@ -0,0 +1,267 @@
+    // if the name already exists, append the field id to the name
+    if uniq_name_set.contains(&field.name) {
+      let new_name = format!("{}-{}", field.name, field.id);
+      field.name.clone_from(&new_name);
+    }
+    uniq_name_set.insert(field.name.clone());
+    field_by_id.insert(field.id.clone(), field);
+  }
+  field_by_id
+}
+
+/// create a map type option writer by field id
+pub fn type_option_writer_by_id(
+  fields: &[Field],
+) -> HashMap<String, Box<dyn TypeOptionCellWriter>> {
+  let mut type_option_reader_by_id: HashMap<String, Box<dyn TypeOptionCellWriter>> =
+    HashMap::with_capacity(fields.len());
+  for field in fields {
+    let field_id: String = field.id.clone();
+    let type_option_reader: Box<dyn TypeOptionCellWriter> = {
+      let field_type: &FieldType = &FieldType::from(field.field_type);
+      let type_option_data: TypeOptionData = match field.type_options.get(&field_type.type_id()) {
+        Some(tod) => tod.clone(),
+        None => HashMap::new(),
+      };
+      type_option_cell_writer(type_option_data, field_type)
+    };
+    type_option_reader_by_id.insert(field_id, type_option_reader);
+  }
+  type_option_reader_by_id
+}
+
+/// create a map type option reader by field id
+pub fn type_option_reader_by_id(
+  fields: &[Field],
+) -> HashMap<String, Box<dyn TypeOptionCellReader>> {
+  let mut type_option_reader_by_id: HashMap<String, Box<dyn TypeOptionCellReader>> =
+    HashMap::with_capacity(fields.len());
+  for field in fields {
+    let field_id: String = field.id.clone();
+    let type_option_reader: Box<dyn TypeOptionCellReader> = {
+      let field_type: &FieldType = &FieldType::from(field.field_type);
+      let type_option_data: TypeOptionData = match field.type_options.get(&field_type.type_id()) {
+        Some(tod) => tod.clone(),
+        None => HashMap::new(),
+      };
+      type_option_cell_reader(type_option_data, field_type)
+    };
+    type_option_reader_by_id.insert(field_id, type_option_reader);
+  }
+  type_option_reader_by_id
+}
+
+pub fn type_options_serde(
+  type_options: &TypeOptions,
+  field_type: &FieldType,
+) -> HashMap<String, serde_json::Value> {
+  let type_option = match type_options.get(&field_type.type_id()) {
+    Some(type_option) => type_option,
+    None => return HashMap::new(),
+  };
+
+  let mut result = HashMap::with_capacity(type_option.len());
+  for (key, value) in type_option {
+    match field_type {
+      FieldType::SingleSelect | FieldType::MultiSelect | FieldType::Media => {
+        if let yrs::Any::String(arc_str) = value {
+          if let Ok(serde_value) = serde_json::from_str::<serde_json::Value>(arc_str) {
+            result.insert(key.clone(), serde_value);
+          }
+        }
+      },
+      _ => {
+        result.insert(key.clone(), serde_json::to_value(value).unwrap_or_default());
+      },
+    }
+  }
+
+  result
+}
+
+pub fn collab_from_doc_state(doc_state: Vec<u8>, object_id: &str) -> Result<Collab, AppError> {
```

Comment:
Duplicated code. Please check biz::collab::ops

Reply:
I simply move this function from there to `utils.rs`. it's not duplicated